javascript/utils.js

```diff
@@ -131,14 +131,26 @@ function handleErrors (response) {
 }
 
 function safeString (str) {
+  if (typeof str !== 'string') {
+    console.warn(`Operation requires a String, but got ${str} (${typeof str})`)
+    return ''
+  }
   return str != null ? String(str) : ''
```

Comment:
No, I don't like it. This means if I transmit my operation that should set an attribute to the integer of 1. It will allow this operation server side, but then on client side will cast it to `''`, which is not what I would expect.

This may actually break existing apps, that transmits Integers to dataset value, where they relied on JS default behavior to cast it to a string.

Considering where my thing came from
```ruby
set_dataset_property(selector: "[data-add-to-cart-item-gid-value=\"#{@purchasable.to_gid_param}\"]",
      name: "addToCartQuantityValue",
      value: quantity,
      select_all: true)
```

where quantity was an integer, it worked fine, except for when it was `0`, now it would fail for all numbers, unless I cast it `to_s` first server side. So I would expect to be automatically cast either server side, when creating the operation or client side.

Reply:
Alright, thanks to quality logic review from @RolandStuder I believe that the correct course of action is to introduce a 4th function, `safeScalar`. It warns if the value passed is not `['string', 'number', 'boolean'].includes(typeof val)` and doesn't mess with the value passed, unless the value passed is null or undefined (which gets corrected to `''`).

I've done a pass through the operations to update them to use `safeString` and `safeScalar` where appropriate. For example, when operating on CSS class names or property names, `safeString` is applied. (CSS classes shouldn't start with numbers, even though many browsers seem to allow it for compatibility.) For the values themselves, `safeScalar`.

The one scenario that is more complex is actually `morph`, since it works on a `template` element. What I did here was replace `String(html).trim()` with `String(safeScalar(html)).trim()` because some folks might conceivably want to send a `0` to morph.

I didn't attempt to analyze or mutate any `url` values eg. `go`, `pushState`, `redirectTo` because I don't want to get into actual validation. Open to opinions here.

Anyhow, what do you think of the solution as presented? Is `string`, `number`, `boolean` the complete list?

Reply:
I like `safeScalar` and think that `string`, `number`, and `boolean` cover our use cases of what will emit from the server. Also, the logic outlined above for URLs etc... seems reasonable to me, but I also think expecting a String or even Scalar for URLs is "good enough" for our needs.

Reply:
I just pushed a commit that protects `url` options with `safeString`.

Reply:
BTW I don't think that a literal number can be interpreted as a valid URL.

Interestingly, even with `666` in my `/etc/hosts`, `ping 666` resolves to `0.0.2.154`. Of course `ping 667` resolves to `0.0.2.155`.

TL;DR: scalar is not necessary.